packages/dds/tree/src/test/feature-libraries/default-schema/schemaChecker.spec.ts

```diff
@@ -0,0 +1,722 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+
+import { Multiplicity } from "../../../../dist/index.js";
+// Reaching into internal module just to test it
+import {
+	SchemaValidationErrors,
+	compliesWithMultiplicity,
+	isFieldInSchema,
+	isNodeInSchema,
+	// eslint-disable-next-line import/no-internal-modules
+} from "../../../feature-libraries/default-schema/schemaChecker.js";
+import {
+	FieldKinds,
+	type FlexFieldKind,
+	type FullSchemaPolicy,
+} from "../../../feature-libraries/index.js";
+import {
+	LeafNodeStoredSchema,
+	MapNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	ValueSchema,
+	type FieldKey,
+	type FieldKindData,
+	type FieldKindIdentifier,
+	type MapTree,
+	type StoredSchemaCollection,
+	type TreeFieldStoredSchema,
+	type TreeNodeSchemaIdentifier,
+	type TreeNodeStoredSchema,
+	type Value,
+} from "../../../core/index.js";
+import { brand } from "../../../util/index.js";
+import type { IFluidHandle, IRequest, IResponse } from "@fluidframework/core-interfaces";
+
+/**
+ * A fake FluidHandle to be used in tests.
+ * It's completely non-functional other than the `IFluidHandle` property, which is required
+ * for the validations that SharedTree does on a FluidHandle value in a node.
+ */
+class TestFluidHandle implements IFluidHandle {
+	public absolutePath: string = "fakePath";
+	public isAttached: boolean = false;
+
+	public get IFluidHandle(): IFluidHandle {
+		return this;
+	}
+
+	public async get(): Promise<any> {
+		throw new Error("Method not implemented.");
+	}
+
+	public bind(handle: IFluidHandle): void {
+		throw new Error("Method not implemented.");
+	}
+
+	public attachGraph(): void {
+		throw new Error("Method not implemented.");
+	}
+
+	public async resolveHandle(request: IRequest): Promise<IResponse> {
+		throw new Error("Method not implemented.");
+	}
+}
+
+const emptySchemaPolicy: FullSchemaPolicy = {
+	fieldKinds: new Map(),
+};
+
+const emptySchemaCollection: StoredSchemaCollection = {
+	nodeSchema: new Map(),
+};
+
+/**
+ * Helper for building {@link TreeFieldStoredSchema}.
+ */
+function getFieldSchema(
+	kind: { identifier: FieldKindIdentifier },
+	allowedTypes?: Iterable<TreeNodeSchemaIdentifier>,
+): TreeFieldStoredSchema {
+	return {
+		kind: kind.identifier,
+		types: allowedTypes === undefined ? undefined : new Set(allowedTypes),
+	};
+}
+
+function getValueNode(nodeSchemaIdentifier: string, value: Value): MapTree {
+	return {
+		type: brand(nodeSchemaIdentifier),
+		value,
+		fields: new Map(),
+	};
+}
+function getMapNode(nodeSchemaIdentifier: string, fields: Map<FieldKey, MapTree[]>): MapTree {
+	return {
+		type: brand(nodeSchemaIdentifier),
+		value: undefined,
+		fields,
+	};
+}
+
+function getObjectNode(nodeSchemaIdentifier: string, fields: Map<FieldKey, MapTree[]>): MapTree {
+	return {
+		type: brand(nodeSchemaIdentifier),
+		value: undefined,
+		fields,
+	};
+}
```

Comment:
As these two are the same, combining them into something like `createNonLeaf` might be nice.